identity_core/src/common/single_struct_error.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::borrow::Cow;
+use std::error::Error;
+use std::fmt::Debug;
+use std::fmt::Display;
+
+/// A container implementing the [`std::error::Error`] trait.
+///
+/// Instances always carry a corresponding `kind` of type `T` and may be extended with custom error
+/// messages, source and recovery data.
+///
+/// This type is mainly designed to accommodate for the [single struct error design pattern](https://nrc.github.io/error-docs/error-design/error-type-design.html#single-struct-style).
+///
+/// When used in a specialized context it is recommended to use a type alias (i.e. `type MyError =
+/// SingleStructError<MyErrorKind>` or `type MyError = SingleStructError<MyErrorKind, MyRecoveryData>`).
+#[derive(Debug)]
+pub struct SingleStructError<T: Debug + Display, S: Debug = ()> {
+  repr: Repr<T, S>,
+}
+
+impl<T: Display + Debug, S: Debug> Display for SingleStructError<T, S> {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    match self.repr {
+      Repr::Simple(ref kind) => write!(f, "{}", kind),
+      Repr::Extensive(ref extensive) => {
+        write!(f, "{}", &extensive.kind)?;
+        let Some(ref message) = extensive.message else {return Ok(())};
+        write!(f, " message: {}", message.as_ref())
+      }
+    }
+  }
+}
+
+impl<T: Debug + Display, S: Debug> Error for SingleStructError<T, S> {
+  fn source(&self) -> Option<&(dyn Error + 'static)> {
+    self.extensive().and_then(|err| {
+      err
+        .source
+        .as_ref()
+        .map(|source| source.as_ref() as &(dyn Error + 'static))
+    })
+  }
+}
+
+#[derive(Debug)]
+struct Extensive<T: Debug + Display, S: Debug> {
+  kind: T,
+  source: Option<Box<dyn Error + Send + Sync + 'static>>,
+  message: Option<Cow<'static, str>>,
+  recovery_data: Option<S>,
+}
+
+#[derive(Debug)]
+enum Repr<T: Debug + Display, S: Debug> {
```

Comment:
This is not changed in this PR, but can we use a better name here? Maybe `ErrorContent`? Or we can remove it completely and just use the `Extensive` variant directly in main struct since it includes the "kind" and all other properties are optional? Yes it needs a heap allocation but that's not too bad?

This is optional, I'm fine with how it is rn.

Reply:
I agree that the name is perhaps not very descriptive, but it is luckily only used internally. Since there is some prior art for this name and design (I essentially copied it from [`std::io::Error`](https://doc.rust-lang.org/src/std/io/error.rs.html#69)) I think it is OK to keep it as is.